src/mmfatl.c

```diff
@@ -10,22 +10,94 @@
  * conditions (corrupt state, out of memory).
  */
 
-#include <stdbool.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 
 #include "mmfatl.h"
 
+#define NUL '\x00'
+
+//----------
+
+/*
+ * During development you may not want to expose preliminary results to the
+ * normal compile, as this would trigger 'unused' warnings, for example.  In
+ * the regression test environment your code may be referenced by testing code,
+ * though.
+ *
+ * This section should be empty, or even removed, once your development is
+ * finished.
+ */
+#if TEST_ENABLE
+#   define UNDER_DEVELOPMENT
+#endif
+
+#ifdef UNDER_DEVELOPMENT
+/* the size a fatal error message including the terminating NUL character can
+ * assume without truncation.
+ */
+#define BUFFERSIZE 1024
+
+/* the character sequence appended to a truncated fatal error message, so a
+ * user is aware a displayed message is incomplete.
+ */
+#define ELLIPSIS "..."
+
+char buffer[BUFFERSIZE + sizeof(ELLIPSIS)];
+
+/*
+ * We do not rely on any initialization during program start.  Instead we
+ * assume the worst case, a corrupted pointer overwrote the buffer.  So we
+ * initialize it again immediately before use.
+ */
+static void initBuffer()
+{
+    char ellipsis[] = ELLIPSIS;
+
+    memset(buffer, NUL, BUFFERSIZE);
+    memcpy(buffer + BUFFERSIZE, ellipsis, sizeof(ellipsis));
+}
+
+#endif // UNDER_DEVELOPMENT
+
 //=================   Regression tests   =====================
 
-#ifdef TEST_ENABLE
+#if TEST_ENABLE
 
 /*  automatic testing to prevent regression   */
 
+bool test_initBuffer(bool silent)
+{
+    // emulate memory corruption
+    memset(buffer, 'x', sizeof(buffer));
+
+    initBuffer();
+
+    bool ok = true;
+    unsigned i = 0;
+    // check the buffer is filled with NUL...
+    for (; ok && i < BUFFERSIZE; ++i)
+        ok = buffer[i] == NUL;
+    // ...and has the ELLIPSIS string at the end...
+    char ellipsis[] = ELLIPSIS;
+    unsigned j = 0;
+    for (; ok && ellipsis[j] != NUL; ++i, ++j)
+        ok = buffer[i] == ellipsis[j];
+    ok = ok && buffer[i] == NUL;
+    if (!ok)
+        printf("initBuffer: initialization failed\n");
+    else if (!silent)
+        printf("initBuffer OK\n");
```

Comment:
This line (and the logic around `silent`) should be in the test harness, not the actual test. I think the test should only be responsible for the failure message and boolean return value, and the harness follows that up with general information about the tests. (You can use a macro to stringify the test name to put it in the printf.) So the harness would look something like:
```
RUN_TEST(test_initBuffer);
RUN_TEST(testSuccessMessage);
```
and there is some code around this and in the macro to handle accumulating an `ok` variable and printing ok/fail stuff.

Reply:
https://en.wikipedia.org/wiki/List_of_unit_testing_frameworks#C

One problem is that C99 predates many of them...
Lets address this step by step.  Again, it is not my primary goal in this series to present a fool-proof test suite, just to give it a head start.

Reply:
I wasn't suggesting using an off the shelf unit testing framework (although the way this is going I wouldn't be surprised if it ends up being a better investment in the end). Here `RUN_TEST(x)` is something like:
```c
#define RUN_TEST(x) \
  if (x()) { printf("test " #x " ok\n"); } else { ok = false; printf("TEST " #x " FAILED\n"); }
```

Reply:
For now I'm trying to keep the code growth as a result of these tests as small as possible, so low-tech and concise is good. We still aren't testing anything that really matters here, like verification and proof assistant activities, so this investment hasn't paid off yet from my perspective.

Reply:
ok, tomorrow.  Its about midnight here.

Reply:
1. It is easier to decide where to refactor and optimize when you have a handful test cases at hand.  It is a bit more difficult to anticipate this in advance.  Or in other words, the code gets better with the experience you have.
2. I caught two bugs with the parser in my original #86 PR.  And I needed to convert the source from int to bool, and the correctness was verified by the test suite.  Not bad, I'd say.